dev-python/sortedcontainers: Version bump 2.1.0

AndrewAmmerlaan opened the pull request:
Breaking up PR into smaller parts: https://github.com/gentoo/gentoo/pull/13988

Bug: https://bugs.gentoo.org/702458
Signed-off-by: Andrew Ammerlaan <andrewammerlaan@riseup.net>
Package-Manager: Portage-2.3.81, Repoman-2.3.20

hartwork commented:
Also: https://bugs.gentoo.org/702590

hartwork commented on dev-python/sortedcontainers/sortedcontainers-2.1.0.ebuild:
```diff
@@ -0,0 +1,22 @@
+
+DESCRIPTION="Python library to sort collections and containers"
+HOMEPAGE="http://www.grantjenks.com/docs/sortedcontainers/"
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
```
The PyPI container seems to lack the test files so let's use the GitHub release tarball instead, please.

hartwork commented on dev-python/sortedcontainers/sortedcontainers-2.1.0.ebuild:
```diff
@@ -0,0 +1,22 @@
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+PYTHON_COMPAT=( python2_7 python3_{6,7,8} )
```
`sortedcontainers-1.5.7.ebuild` has Python 3.5 support enabled and upstream hasn't dropped support for 3.5. according to `setup.py` metadata. Why has 3.5 been dropped during the bump?

AndrewAmmerlaan replied:
@flappyports told me it could be dropped here: https://github.com/gentoo/gentoo/pull/13988

>py3_5 can be dropped as it will be dropped from Gentoo soon.

hartwork replied:
Dropping Python 3.5 will need touching many ebuilds at once: a mass-removal. Before that happens, I don't see how thinning support out at some place helps the big picture unless python support 3.5 is known broken in that package. Am I missing something? @flappyports what do you think?

hartwork commented on dev-python/sortedcontainers/sortedcontainers-2.1.0.ebuild:
```diff
@@ -0,0 +1,22 @@
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
```
Why drop the existing `~` arches here? It doesn't seem necessary.

AndrewAmmerlaan replied:
I dropped them because I can't test the package on these architectures, and as I understood it, I shouldn't include architectures I can't test on. But I can re-add them I suppose.

hartwork replied:
For new packages: yes! For this case with a bump — assuming no known breakages — I believe it's safe to keep in. Happy to learn otherwise.

hartwork commented on dev-python/sortedcontainers/sortedcontainers-2.1.0.ebuild:
```diff
@@ -0,0 +1,22 @@
+DEPEND="${PYTHON_DEPS}
+	dev-python/setuptools[${PYTHON_USEDEP}]"
+
+RDEPEND="${DEPEND}"
```
I'd expect empty…
```
DEPEND=
RDEPEND=
```
…here and have the `distutils-r1` eclass pull in everything else — including setuptools and pytest — but I'll need to check again. 
For `PYTHON_DEPS` please drop and check https://wiki.gentoo.org/wiki/Project:Python/distutils-r1#PYTHON_DEPS .

AndrewAmmerlaan replied:
Are you sure about setuptools? I see it specified explicitly in a lot of packages in the dev-python category.

I'll drop the `PYTHON_DEPS` in the next commit

hartwork replied:
Not 100% sure yet. I believe I saw something like it in the eclass. I'm running out of time for an in-depth look right now.

AndrewAmmerlaan replied:
I've been reading [this wiki page](https://wiki.gentoo.org/wiki/Project:Python/Dependencies#Reading_setuptools_setup.py_files) and if I understand it correctly both setuptools and `PYTHON_DEPS` should be specified:

> PYTHON_DEPS includes the dependency on the Python interpreter and dev-lang/python-exec. If the package installs Python scripts, modules or links to Python libraries, it needs to include PYTHON_DEPS in RDEPEND. In other words, every ebuild using distutils-r1, python-r1, or python-single-r1 needs to use it. 

> The easy rule is: always include PYTHON_DEPS in DEPEND. 

> Setuptools is an extension to the standard distutils build system commonly used by Python packages. In order to determine whether the package uses setuptools, you should check the setup.py file. If the setuptools module (or any function from it) is imported, then dev-python/setuptools needs to be specified as a build-time dependency.

hartwork replied:
I like where this is going — interesting!

This is what empty `DEPEND` and `RDEPEND` in the ebuild get me with `distutils-r1`:
- `BDEPEND`: `test? ( pytest )` + `PYTHON_DEPS` + setuptools
- `DEPEND`: empty
- `RDEPEND`: `PYTHON_DEPS`

Unless `setuptools` is a true runtime dependency for this package — I find no imports to `setuptools` or `pkg_resources` outside of `setup.py` and no entrypoints in `setup.py` — that doesn't look too bad to me.

pytest was put in by `distutils_enable_tests pytest`:
https://github.com/gentoo/gentoo/blob/d51fec51c03a67074cf521d274e47ebcf1e1c412/eclass/distutils-r1.eclass#L396-L401

setuptools is in there because of default `${DISTUTILS_USE_SETUPTOOLS:=bdepend}` and its effect to `_distutils_set_globals`:
https://github.com/gentoo/gentoo/blob/cbb12d98d9e63b7e5f447aa170412a85550da5b4/eclass/distutils-r1.eclass#L118-L127

This is what I used as a quickhack to get at the final variable values:
```console
# ebuild sortedcontainers-2.1.0.ebuild manifest clean unpack
# for i in BDEPEND DEPEND RDEPEND ; do echo "[$i]"; ( source /var/tmp/portage/dev-python/sortedcontainers-2.1.0/temp/environment ; eval "echo \${$i}" ) | sed 's, ,\n,g' | sed 's,^,  ,' ; echo ; done
```

> The easy rule is: always include PYTHON_DEPS in DEPEND.

I think that's from a time before EAPI 7 was mainstream because `distutils-r1` today adds `PYTHON_DEPS` to [`BDEPEND`](https://dev.gentoo.org/~mgorny/articles/the-ultimate-guide-to-eapi-7.html#build-time-dependencies-split-into-depend-and-bdepend) from EAPI 7 on (and to `DEPEND` before). To prove that with code:
https://github.com/gentoo/gentoo/blob/d51fec51c03a67074cf521d274e47ebcf1e1c412/eclass/distutils-r1.eclass#L134-L138

Since that wiki page is not specific to `distutils-r1`, maybe that advice was mean to be on the safe side even with ebuilds where only a small piece of the software uses Python — just guessing.
My source would be https://wiki.gentoo.org/wiki/Project:Python/distutils-r1#PYTHON_DEPS:
> [PYTHON_DEPS] should be used only with DISTUTILS_OPTIONAL enabled.


What do you think?

AndrewAmmerlaan commented:
> This is what I found:

@hartwork I have updated the ebuild with your suggestions, I changed the SRC_URI, re-added python3_5 (it is only scheduled to be dropped somewhere at the end of next year), re-added the arches and removed the unnecessary DEPEND, and RDEPEND.

hartwork commented:
Related: eaf356158863f900c04362bc531624cb5f1f27ca

hartwork commented on dev-python/sortedcontainers/sortedcontainers-2.1.0.ebuild:
```diff
@@ -0,0 +1,17 @@
+
+DESCRIPTION="Python library to sort collections and containers"
+HOMEPAGE="http://www.grantjenks.com/docs/sortedcontainers/"
+SRC_URI="https://github.com/grantjenks/python-${PN}/archive/v${PV}.tar.gz"
```
(Additional ` -> ${P}.tar.gz` would have been needed here or the filename would be `v2.1.0.tar.gz` which is not ideal in a single folder on a mirror.)

AndrewAmmerlaan replied:
Aah yes, I see, that would indeed have been better. I will take that into account in future ebuilds.
